### DISCUSSION TITLE
How bootstrap-container's user-data works?

### DISCUSSION THREAD

**Discussion by wzhiana** [bottlerocket-os/bottlerocket#3728] [General]
Hello, 

I'm trying to use aws-cli image. As input it waits some command for it. 

and i tried to put encoded help (`aGVscAo=`), and expected to see message, which i can get locally by: 
```
docker run amazon/aws-cli:latest help
```

config:
```
[settings.bootstrap-containers.aws]
              source = "amazon/aws-cli:latest"
              mode = "once"
              essential = false
              user-data = "aGVscAo="
```

but all i get - is the error:
```
Jan 24 12:47:42 ip-10-216-35-53.eu-central-1.compute.internal host-ctr[1624]: time="2024-01-24T12:47:42Z" level=info msg="unpacking image..." img="amazon/aws-cli:latest"
Jan 24 12:47:49 ip-10-216-35-53.eu-central-1.compute.internal host-ctr[1624]: time="2024-01-24T12:47:49Z" level=info msg="Container does not exist, proceeding to create it" ctr-id=boot.ssm
Jan 24 12:47:50 ip-10-216-35-53.eu-central-1.compute.internal host-ctr[1624]: time="2024-01-24T12:47:50Z" level=info msg="container task does not exist, proceeding to create it" container-id=boot.ssm
Jan 24 12:47:50 ip-10-216-35-53.eu-central-1.compute.internal host-ctr[1624]: time="2024-01-24T12:47:50Z" level=info msg="successfully started container task"
Jan 24 12:47:50 ip-10-216-35-53.eu-central-1.compute.internal host-ctr[1624]: usage: aws [options] <command> <subcommand> [<subcommand> ...] [parameters]
Jan 24 12:47:50 ip-10-216-35-53.eu-central-1.compute.internal host-ctr[1624]: To see help text, you can run:
Jan 24 12:47:50 ip-10-216-35-53.eu-central-1.compute.internal host-ctr[1624]:   aws help
Jan 24 12:47:50 ip-10-216-35-53.eu-central-1.compute.internal host-ctr[1624]:   aws <command> help
Jan 24 12:47:50 ip-10-216-35-53.eu-central-1.compute.internal host-ctr[1624]:   aws <command> <subcommand> help
Jan 24 12:47:50 ip-10-216-35-53.eu-central-1.compute.internal host-ctr[1624]: aws: error: the following arguments are required: command
```

so, how user-data is being passed?

**Comment by gthao313** [comment]
@yaroslav-nakonechnikov Thanks for starting this conversation. 

After digging into the aws-cli image, I don't think aws-cli container has the mechanism to support you to input user-data there. More useful info [here](https://github.com/aws/aws-cli/blob/v2/docker/Dockerfile), Which means you are unable to config the user-data  on aws-cli container. 

If you need to use a container that knows how to read user-data, I would suggest your to use the container like AmazonLinux container : ). 

Please let me know if you have more questions. thanks!

**Reply by wzhiana** [reply]
yes, i saw docker file. But it doesn't give me details. 

i want to know how that user-data is being passed to container. If it should be just string as it is written, so just `help` should provide correct output in the logs and exit code is 0. 
But it is not like that.

**Reply by wzhiana** [reply]
i found one example: https://github.com/stefansundin/bottlerocket-bootstrap-associate-eip 
but it also looks like passing string. and it also have entrypoint and nothing which is called user-data

**Reply by gthao313** [reply]
I think you could do the above way. The example you linked has a rust tool reading the user-data -> https://github.com/stefansundin/bottlerocket-bootstrap-associate-eip/blob/main/src/main.rs.

FYI, we have new guides here. https://bottlerocket.dev/en/os/1.18.x/concepts/bootstrap-containers/ and  https://bottlerocket.dev/en/os/1.18.x/api/settings/bootstrap-containers/.

**Reply by wzhiana** [reply]
so, that user-data is a file, which should be read additionally? 
and entrypoint should look like: `bash /.bottlerocket/bootstrap-containers/current/user-data`  (in case user data is bash script)?

**Reply by gthao313** [reply]
I think you are correct. I just tried on my end and the `bash ./user-data` did the right thing without setting the exec flag in the user data file.

Meanwhile, we have an opening issue that takes the user data and execute it as a shell script. Feel free to left comments there!
https://github.com/bottlerocket-os/bottlerocket/issues/2010
